### PR TITLE
ChatService & Case Owner Assignment & Case Sharing.

### PR DESCRIPTION
**Pull request by vkkurre**
1. Chat Transcript will be inserted as Case Comment. - Flow
2. Case Owner Assignment to one of the agent when community user create a case from community or anywhere. - Apex Class
3. Case Created by Chat Service by community user should be shared to community user. - Apex Class & Apex Trigger.
4. Custom Notification should be thrown to the user ( either agent or customer ) while making case comment.  - Flow

### CODE REVIEW

File: Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/*****************************************************************************************************************
+* 
+* @ Name            :   DEL_CaseTriggerHandler
+* @ Purpose         :   Controller for Case to assign Case Owner and sharing Case to certain User.
+* @ Author          :   Vinay kant
+* @ Usage           :   1) To Assigning a case to one of the available service agent user.
+*                       2) To Share the case record by using Apex Sharing to user created that case from Chat.
+* @ Test Class Name :
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls:8]: Add test class name

File: Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+* @ Name            :   DEL_CaseTriggerHandler
+* @ Purpose         :   Controller for Case to assign Case Owner and sharing Case to certain User.
+* @ Author          :   Vinay kant
+* @ Usage           :   1) To Assigning a case to one of the available service agent user.
+*                       2) To Share the case record by using Apex Sharing to user created that case from Chat.
+* @ Test Class Name :   
+*
+*   Date            |  Developer Name               |  Version      |  Changes
+* =================================================================================================================
+*   07-07-2022      |  vinaykant.kurre@absyz.com      |  1.0          |  Initial Version
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls:12]: Align the lines rightly

File: Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+*   
+*****************************************************************************************************************/
+
+public with sharing class DEL_CaseTriggerHandler {
+    public static Boolean blnSkipTrigger = false;
+    
+    public class ReturnValuesHandler{
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls:19]:
```suggestion
    public class ReturnValuesHandler {
```

File: Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        @AuraEnabled
+        public Boolean blnIsSuccess;
+        @AuraEnabled 
+        public String  strErrorMessage;
+    }
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls:24]: Maintain a line of gap after every logic/block

File: Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+/**
+    * @ author      : Vinaykant
+    * @ description : Invocable Method for Flow to assign case owner to one of the Service Agent User.
+    * @ params      : 'list_users' - List of Users in the queue fetched by flow.
+    * @ return      : 'list_SelectedUserIds' - List of Selected User Id from queue (i.e. One User Id for Flow ).
+**/
+    @InvocableMethod(label = 'Assign Case Owner' 
+                    description = 'Assigning Case Owner to newly created Case.')
+    public static List<Id> assigningCaseOwner(List<List<GroupMember>> list_users){
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls:33]:
```suggestion
    public static List<Id> assigningCaseOwner(List<List<GroupMember>> list_Users) {
```

File: Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        List<Id> list_SelectedUserIds = new List<Id>();
+        List<GroupMember> list_allGroupUsers = list_users.get(0);
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls:35]: Always have a IF condition checking the list is not empty before using the hard-coded index element 0
```suggestion
        List<GroupMember> list_GroupMembers = list_Users.get(0);
```

File: Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        Integer intTotalUsers = list_allGroupUsers.size();
+        Integer intIndex = (Integer)Math.floor(Math.random() * intTotalUsers);
+        list_SelectedUserIds.add(list_allGroupUsers[intIndex].UserOrGroupId);
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls:38]: Not understanding what are you doing here? What is this used for?

File: Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        return list_SelectedUserIds;
+    }
+
+/**
+    * @ author      : Vinaykant
+    * @ description : Sharing Case to Community User who created Case from Chat Conversation. 
+                      This Method is called by Trigger named 'DEL_CaseSharingTrigger'.
+    * @ params      : 'objCase' - Case Object Record.
+**/
+    public static ReturnValuesHandler sharingCaseToUser(Case objCase){
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls:48]:
```suggestion
    public static ReturnValuesHandler sharingCaseToUser(Case objCase) {
```

File: Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls [hunk 9]
```diff
@@ -50,22 +70,101 @@ public class DEL_CommentsPageController {
...
+        * @ description : Apex Method to delete a particular case comment from a case.
+        * @ params      : 'idCommentId' - Case Id of which all case comments will be fetched.
+        * @ return      : 'objDMLHandler' - Wrapper Class Object.
+    **/
+    @AuraEnabled
+    public static DEL_DMLHandler deleteCaseComment(Id idCommentId){
         DEL_DMLHandler objDMLHandler = new DEL_DMLHandler();
-            objDMLHandler.blnIsSuccess = true;
-            objDMLHandler.strErrorMessage = ' ';
+        DEL_CaseComment__c objCaseComment = [SELECT
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls:102]: SOQL indentation is not right

File: Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls [hunk 10]
```diff
@@ -50,22 +70,101 @@ public class DEL_CommentsPageController {
...
+    @AuraEnabled
+    public static DEL_DMLHandler deleteCaseComment(Id idCommentId){
         DEL_DMLHandler objDMLHandler = new DEL_DMLHandler();
-            objDMLHandler.blnIsSuccess = true;
-            objDMLHandler.strErrorMessage = ' ';
+        DEL_CaseComment__c objCaseComment = [SELECT 
+                                            Id 
+                                            FROM DEL_CaseComment__c 
+                                            WHERE Id = :idCommentId];
+        try{
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls:106]:
```suggestion
        try {
```

File: Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls [hunk 11]
```diff
@@ -50,22 +70,101 @@ public class DEL_CommentsPageController {
...
+            delete objCaseComment;
+        }
+        catch(DMLException objException){
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls:109]:
```suggestion
        catch (DMLException objException) {
```

File: Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls [hunk 12]
```diff
@@ -50,22 +70,101 @@ public class DEL_CommentsPageController {
...
+
+    /**
+        * @ author      : Vinaykant
+        * @ description : Apex Method to update a particular case comment with new edited comment.
+        * @ params      : 'idCommentId' - Case Id of which all case comments will be fetched.
+        *                 'strComment' - New Edited Comment Body String for Case Comment Record.
+        * @ return      : 'objDMLHandler' - Wrapper Class Object.
+    **/
+    @AuraEnabled
+    public static DEL_DMLHandler updateCaseComment(Id idCommentId, String strComment){
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls:129]:
```suggestion
    public static DEL_DMLHandler updateCaseComment(Id idCommentId, String strComment) {
```

File: Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls [hunk 13]
```diff
@@ -50,22 +70,101 @@ public class DEL_CommentsPageController {
...
+        DEL_DMLHandler objDMLHandler = new DEL_DMLHandler();
+        DEL_CaseComment__c objCaseComment = [SELECT
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/classes/DEL_CommentsPageController.cls:131]: SOQL indentation is not right

File: Delmar/force-app/main/default/flows/DEL_Case_Comment_Flow.flow-meta.xml [hunk 14]
```diff
@@ -0,0 +1,335 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/flows/DEL_Case_Comment_Flow.flow-meta.xml:1]: Name of the Flow is not right, it is not describing what the flow is used for

File: Delmar/force-app/main/default/lwc/dEL_deleteWizard/dEL_deleteWizard.html [hunk 15]
```diff
@@ -0,0 +1,35 @@
+<template>
+<!--
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/lwc/dEL_deleteWizard/dEL_deleteWizard.html:2]: File comment should be the first, before template

File: Delmar/force-app/main/default/lwc/dEL_deleteWizard/dEL_deleteWizard.js [hunk 16]
```diff
@@ -0,0 +1,24 @@
+import { LightningElement,api } from 'lwc';
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/lwc/dEL_deleteWizard/dEL_deleteWizard.js:1]:
```suggestion
import { LightningElement, api } from 'lwc';
```

File: Delmar/force-app/main/default/lwc/dEL_deleteWizard/dEL_deleteWizard.js [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    handleCloseButton(){
+        this.dispatchEvent(new CustomEvent('close'));
+    }
+    
+/**
+    * @ author      : Vinaykant
+    * @ description : Creating a custom event for deleting the current Case Comment.
+**/
+    handleSaveButton(){
+        let saveEvent = new CustomEvent('save',{detail:this.strCommentId})
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/lwc/dEL_deleteWizard/dEL_deleteWizard.js:20]:
```suggestion
        let saveEvent = new CustomEvent('save', { detail:this.strCommentId })
```

File: Delmar/force-app/main/default/lwc/del_EditWizard/del_EditWizard.js-meta.xml [hunk 18]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>54.0</apiVersion>
+    <isExposed>true</isExposed>
+    <targets>
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/lwc/del_EditWizard/del_EditWizard.js-meta.xml:5]: I dont see the need for these target configs. This component will be used inside another component right?

File: Delmar/force-app/main/default/lwc/del_commentsPage/del_commentsPage.js [hunk 19]
```diff
@@ -1,31 +1,95 @@
 import { LightningElement,api,track, wire } from 'lwc';
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/lwc/del_commentsPage/del_commentsPage.js:1]:
```suggestion
import { LightningElement, api, track, wire } from 'lwc';
```

File: Delmar/force-app/main/default/lwc/del_commentsPage/del_commentsPage.js [hunk 20]
```diff
@@ -1,31 +1,95 @@
...
-
+import Id from '@salesforce/user/Id';
 
 export default class Del_commentsPage extends NavigationMixin(LightningElement) {
     strBody = "";
     @api recordId;
     @track list_Comments = [];
     @track blnIsLoading = false;
-    strInsertId;
+    @track strCurrentUserUrl;
```
Comment (pradeepchary) [inline Delmar/force-app/main/default/lwc/del_commentsPage/del_commentsPage.js:18]: tarck parameter is not needed for PRIMITIVE variables

File: Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls (hunk 7 above)
Comment (vkkurre) [reply Delmar/force-app/main/default/classes/DEL_CaseTriggerHandler.cls:38]: As we have discussed earlier that from community site I can't be able to trigger Case Assignment Rules, So I have written this logic to assign one of the user a Case Owner. This method is returning one user Id from the Queue to the flow. As this invocable method is being called from flow. And Also we have discussed that I and Ankit will be looking into this matter.

File: Delmar/force-app/main/default/lwc/del_EditWizard/del_EditWizard.js-meta.xml (hunk 18 above)
Comment (vkkurre) [reply Delmar/force-app/main/default/lwc/del_EditWizard/del_EditWizard.js-meta.xml:5]: Okay I will Update the target Configs.

File: Delmar/force-app/main/default/lwc/del_commentsPage/del_commentsPage.js (hunk 20 above)
Comment (vkkurre) [reply Delmar/force-app/main/default/lwc/del_commentsPage/del_commentsPage.js:18]: We removed this Variable. strInsertId - Not required.

**Pull request closed without merging**